keps/sig-network/3726-standard-application-protocols/README.md

```diff
@@ -0,0 +1,276 @@
+* While in Istio it will be h2c (HTTP2 over cleartext).
+
+That creates a problem where users with GKE and Istio in their cluster can have very different behaviors for the same `appProtocol` value.
+
+
+
+### Goals
+
+* Build consensus around how common (non IANA service names) should be implemented
+* Help the broader community and specifically implementations to interoperate better
+* Provide short and clear documentation for how appProtocol values should be implemented
+  * Update the appProtocol user documentation respectively
+
+
+### Non-Goals
+
+* Validate appProtocol values
+* Monitor appProtocol implementations
+
+
+## Proposal
+
+Kubernetes supports the `appProtocol` field to provide a way to specify an application protocol for each Service port.
+
+The [documentation](https://kubernetes.io/docs/concepts/services-networking/service/#application-protocol) for this field says that:
+
+```Values should either be IANA standard service names or domain prefixed names such as mycompany.com/my-custom-protocol.```
+
+This KEP proposes to declare standard Kubernetes protocol names for common protocols that are not IANA standard service names.
+
+Those common protocols will be well defined strings prefixed with ‘k8s.io’. 
+`k8s.io/h2c` as an example.
+
+### list of protocols
+- 'k8s.io/h2c'
+- 'k8s.io/h2'
+- 'k8s.io/grpc'
+- 'k8s.io/grpc-web'
+- 'k8s.io/tls'
+- 'k8s.io/tcp'
+
+### Risks and Mitigations
+
+There are no real “risks”, primary concerns are:
+1. End users will not migrate to `k8s.io/<>` values
+2. It will take long time to implementations align with the new standards
+3. We have no easy way to monitor who is aligned and who is not
+
+
+## Design Details
+
+At first, the list is going to live in `ServicePort` and `EndpointPort` as part of the AppProtocol description.
+
+Based on the list size, we might revisit this decision in the future and suggest an alternative location.
+
+Proposed changes to the API spec:
+
+```go
+type ServicePort struct {
+  ... 
+  ...
+
+  // The application protocol for this port.
+  // This field follows standard Kubernetes label syntax.
+  // Valid values are either:
+  //
+  // * Un-prefixed protocol names - reserved for IANA standard service names (as per
+  // RFC-6335 and https://www.iana.org/assignments/service-names).
+  //
+  // * Kubernetes standard names:
+  //   * 'k8s.io/h2' - http2 over TLS 
+  //   * 'k8s.io/h2c' - http2 over cleartext TCP 
+  //   * 'k8s.io/grpc' - grpc traffic
+  //
+  // * Other protocols should use prefixed names such as
+  // mycompany.com/my-custom-protocol.
+  // +optional
+  AppProtocol *string
+```
+
+same wording for type `EndpointPort`
+
+### Adding new protocols
+
+In order to be included in the list, a new protocol must:
+* Not be an [IANA standard service name](https://www.iana.org/assignments/service-names)
+* Run on top of L4 protocol supported by Kubernetes Service
+* Be supported in more than three implementations
+* Be well defined and broadly used
```

Comment:
So this does nothing to stop us from running into the "gke `HTTP2` vs Istio `HTTP2`" problem for new protocols in the future; we have to get three possibly-incompatible implementations first before we're allowed to fix things...

Adding `k8s.io/h2` and `k8s.io/h2c` solves the _specific_ problem that led to this KEP, but we're still leaving the foot gun fully armed for future users.

Reply:
If implementations follow the docs they should not support unprefixed protocols that are not IANA service names. That means that should do "istio.io/<protocol>" or gke.io/<protocol> if the protocol is not IANA standard name. 

I agree that 3 may not be ideal, also the domain prefixed protocol make it impossible for implementations to coexist and support not IANA standard name. For that @robscott suggested future work to turn this field to a list.

When this happens, implementations could truly interoperate. However the field name is no longer appropriate (it is debatable if it was before) since it is not longer the "application protocol" it is more of a hint to the implementations on how to forward the traffic to the application.

Reply:
Yeah I think the ultimate solution is going to be a list that supports multiple values here since there's probably always going to be some implementation specific values.

Reply:
Would it be sufficient to say "at least two in-progress implementations of components which would use this protocol" ?

Reply:
changed it to two or more. 
Does "in progress" is enough for us? I think we'd want to add another standard protocol only when we know there is a real need for it and it is fully supported

Reply:
Yeah, though until it is a list I see Tim's and Dan's point where users will use istio.io/custom or something else and then they would need to migrate to k8s.io
And implementation will need to support both for backward compatibility.

Reply:
if we define the field as an url we can also apply url rules to implement compatibility.  
We can use subdomains to define implementations, and paths to define protocols , it can also allow to do composition for protocols that tunnels  `<implementation>.k8s.io/<protocol1>/<protocol>`

Per example 

```
istio.k8s.io/http/v2
```

will mean the http v2 implementation from istio, when the standard one will be `k8s.io/http/v2`

Something more complex

```
gateway.k8s.io/websocket/sip
```

that means the implementation of SIP over Websocket (I just learned about this recently, no idea it was a thing :)

Reply:
Would Istio and GKE even qualify for the "two implementations" rule?  They used the same name but for different meanings.  If they had both followed the rules and used their own prefixes, we would only have 1 implementation of h2c and one of h2, right?

Perhaps add this as an UNRESOLVED and let's get this merged so we can debate the fine points.

Reply:
Agreed, I have added this as UNRESOLVED.